packages/react/src/metafield-parser.test.ts

```diff
@@ -0,0 +1,539 @@
+    });
+
+    it(`date`, () => {
+      const dateStamp = '2022-10-13';
+      const parsed = metafieldParser<ParsedMetafields['date']>({
+        type: 'date',
+        value: dateStamp,
+      });
+      expect(
+        parsed?.parsedValue?.toString() === new Date(dateStamp).toString()
+      ).toBe(true);
+      expectType<null | Date>(parsed?.parsedValue);
+    });
+
+    it(`date_time`, () => {
+      const dateStamp = '2022-10-13';
+      const parsed = metafieldParser<ParsedMetafields['date_time']>({
+        type: 'date_time',
+        value: dateStamp,
+      });
+      expect(
+        parsed?.parsedValue?.toString() === new Date(dateStamp).toString()
+      ).toBe(true);
+      expectType<null | Date>(parsed?.parsedValue);
+    });
+
+    it(`dimension`, () => {
+      const parsed = metafieldParser<ParsedMetafields['dimension']>({
+        type: 'dimension',
+        value: JSON.stringify({unit: 'mm', value: 2}),
+      });
+      expect(parsed?.parsedValue?.unit === 'mm').toBe(true);
+      expect(parsed?.parsedValue?.value === 2).toBe(true);
+      expectType<null | Measurement>(parsed?.parsedValue);
+    });
+
+    it(`file_reference`, () => {
+      const parsed = metafieldParser<ParsedMetafields['file_reference']>({
+        type: 'file_reference',
+        reference: {
+          __typename: 'GenericFile',
+        },
+      });
+      expect(parsed.parsedValue?.__typename === 'GenericFile').toBe(true);
+      expectType<null | GenericFile>(parsed?.parsedValue);
+    });
+
+    it(`json`, () => {
+      type MyJson = {
+        test: string;
+        bool: boolean;
+        deep: {
+          numb: number;
+        };
+      };
+
+      const myJson = {
+        type: 'json',
+        value: JSON.stringify({test: 'testing', bool: false, deep: {numb: 7}}),
+      };
+
+      // without an extra generic, we just mark it as "unknown"
+      const parsed = metafieldParser<ParsedMetafields['json']>(myJson);
+      // note that with "unknown", you have to cast it as something
+      expect((parsed?.parsedValue as {test: string})?.test === 'testing').toBe(
+        true
+      );
+      expectType<null | unknown>(parsed?.parsedValue);
+
+      // with an extra generic, we use can use that as the type instead
+      const parsedOtherType =
+        metafieldParser<ParsedMetafields<MyJson>['json']>(myJson);
+      expect(parsedOtherType.type === 'json').toBe(true);
+      expect(parsedOtherType.parsedValue?.test === 'testing').toBe(true);
+      expect(parsedOtherType.parsedValue?.bool === false).toBe(true);
+      expect(parsedOtherType.parsedValue?.deep?.numb === 7).toBe(true);
+      expectType<null | MyJson>(parsedOtherType.parsedValue);
+    });
+
+    it(`money`, () => {
+      const parsed = metafieldParser<ParsedMetafields['money']>({
+        type: 'money',
+        value: JSON.stringify({amount: '12', currencyCode: 'USD'}),
+      });
+      // TODO: amount should be a number, not a string
```

Comment:
🤔  I'm wondering about this. If I remember correctly, strings are a way to avoid numerical precision errors.